cuda_core/cuda/core/experimental/_memory.py

```diff
@@ -293,3 +293,30 @@ def is_host_accessible(self) -> bool:
     @property
     def device_id(self) -> int:
         raise RuntimeError("the pinned memory resource is not bound to any GPU")
+
+
+class _SynchronousMemoryResource(MemoryResource):
+    __slots__ = ("_dev_id",)
+
+    def __init__(self, dev_id):
+        self._handle = None
+        self._dev_id = dev_id
+
+    def allocate(self, size, stream=None) -> Buffer:
+        ptr = handle_return(cuda.cuMemAlloc(size))
+        return Buffer(ptr, size, self)
+
+    def deallocate(self, ptr, size, stream=None):
+        handle_return(cuda.cuMemFree(ptr))
```

Comment:
This seems to be out of line with the programming model and what we have done for the others (no implicit sync before freeing). Can you explain which semantic you are referring to please

Reply:
`cudaFree` internally does an expansive synchronization so that users can write this (technically unsafe) code since early days of CUDA:
```c++
int* ptr;
cudaMalloc(&ptr, size);
...
kernel<<<grid, block, shmem, stream>>>(ptr, ...);
cudaFree(ptr);
```
While it is unlikely that `cudaFree` would change this implementation detail, it is not guaranteed by the programming model and an explicit sync before `cudaFree` is encouraged, which brings us to the second point: The same user code based on `cuda.core` should function regardless of TCC/WDDM (or Linux) unless there's any fundamental limitation (ex: lack of HW/FW support). This also aligns with our other memory resources in that the `deallocate` call safe and is ordered on `stream`.

FYI, I fixed a typo in my above suggestion: `stream.sync()` is now called unconditionally.